source/common/request_info/dynamic_metadata_impl.cc

```diff
@@ -0,0 +1,53 @@
+#include "common/request_info/dynamic_metadata_impl.h"
+
+#include "envoy/common/exception.h"
+
+namespace Envoy {
+namespace RequestInfo {
+
+DynamicMetadataImpl::~DynamicMetadataImpl() {
+  for (auto& it : data_storage_) {
+    if (it.second.destructor_) {
+      it.second.destructor_(it.second.ptr_);
+    }
+  }
+  data_storage_.clear();
+}
+
+void DynamicMetadataImpl::setDataGeneric(absl::string_view data_name, size_t type_id, void* data,
+                                         void (*destructor)(void*)) {
+  if (data_storage_.find(data_name) != data_storage_.end()) {
+    (*destructor)(data);
+    throw EnvoyException("DynamicMetadata::setData<T> called twice with same name.");
+  }
+
+  data_storage_[static_cast<std::string>(data_name)] = {type_id, data, destructor};
```

Comment:
Why not combine data name + type ID in the key, so that different types have independent namespaces?

Reply:
So I'm inclined to believe that, philosophically, having a single text based namespace is the right thing.  The concrete example is that if someone tries to access data in a UDH that they set in a filter but didn't use StringAccessor, they'll get a "wrong type" error rather than a "data not found" error.  And I'm having a hard time coming up with an example where I'd want two pieces of data with different types to have the same reverse DNS name; when would that make sense?  Maybe if they were two different ways of saying the same thing?  But in that case, why not have a single class with two different accessors?  (Other than UDH, which needs StringAccessor.)

Reply:
As discussed in person, I agree; we're essentially getting dynamic type checks here, which is a good thing.